src/compiler/checker.ts

```diff
@@ -22555,6 +22555,10 @@ namespace ts {
             return type.flags & TypeFlags.Never || f(type) ? type : neverType;
         }
 
+        function removeType(type: Type, targetType: Type) {
+            return filterType(type, t => t !== targetType);

```

Comment:
This is maybe a micro-optimization, but because unions are sorted, and all the `undefined` type variants are such low type IDs, if `targetType` is a union, we only need to check the first union element, rather than all of them exhaustively, IIRC.

Reply:
Yeah, but `filterType` also does stuff for the (denormalized) origin type and we'd have to duplicate that as well.